### PR TITLE
Improve Job parallel processing expansion task

### PR DESCRIPTION
**Pull request by sftim**
- convert from Concept template to Task template
- explain use of curl for downloading
- use DBPedia URLs for fruit (these should stay valid)
- reword prerequisites
- fix broken link

[preview](https://deploy-preview-17688--kubernetes-io-master-staging.netlify.com/docs/tasks/job/parallel-processing-expansion/)

### CODE REVIEW

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 1]
```diff
@@ -1,50 +1,62 @@
...
 weight: 20
 ---
 
 {{% capture overview %}}
 
-In this example, we will run multiple Kubernetes Jobs created from
-a common template.  You may want to be familiar with the basic,
-non-parallel, use of [Jobs](/docs/concepts/workloads/controllers/jobs-run-to-completion/) first.
+In this example, you will run multiple {{< glossary_tooltip text="Jobs" term_id="job" >}}
+based a common template. You can use this approach to process batches of work in
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md]: @sftim, couple of nits/how about this wording to get rid of _will_:
This example shows/explains how to run multiple Jobs [tooltip stuff] based on a common template.
You can use this approach ...

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 2]
```diff
@@ -1,50 +1,62 @@
...
-{{% capture body %}}
+You should be familiar with the basic,
+non-parallel, use of [Job](/docs/concepts/jobs/run-to-completion-finite-workloads/).
 
-## Basic Template Expansion
+{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}
 
-First, download the following template of a job to a file called `job-tmpl.yaml`
+You will need the command-line utility `sed`, a working install of
+[Python](https://www.python.org/), and the Jinja2 template
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md]: Possibly make the prereqs into a list?

Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: Will that work OK with the `{{< include "task-tutorial-prereqs.md" >}}` and `{{< version-check >}}` shortcodes?

:thinking:

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 1 above)
Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: Reworded, inspired by above suggestion

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 2 above)
Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: How's this?

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 3]
```diff
@@ -1,50 +1,62 @@
...
 weight: 20
 ---
 
 {{% capture overview %}}
 
-In this example, we will run multiple Kubernetes Jobs created from
-a common template.  You may want to be familiar with the basic,
-non-parallel, use of [Jobs](/docs/concepts/workloads/controllers/jobs-run-to-completion/) first.
+This task demonstrates running multiple {{< glossary_tooltip text="Jobs" term_id="job" >}}
+based a common template. You can use this approach to process batches of work in
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md]: line 10, nit, possible edit: s/based/based on/

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 2 above)
Comment (kbhawkey) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: OK

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 4]
```diff
@@ -1,50 +1,62 @@
...
-Unlike a *pod template*, our *job template* is not a Kubernetes API type.  It is just
-a yaml representation of a Job object that has some placeholders that need to be filled
-in before it can be used.  The `$ITEM` syntax is not meaningful to Kubernetes.
+{{% capture steps %}}
 
-In this example, the only processing the container does is to `echo` a string and sleep for a bit.
-In a real use case, the processing would be some substantial computation, such as rendering a frame
-of a movie, or processing a range of rows in a database.  The `$ITEM` parameter would specify for
-example, the frame number or the row range.
+## Create a template Job
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md]: Possible edit:
```##Create a template for the Job```

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 5]
```diff
@@ -180,46 +200,93 @@ spec:
...
+You can pick your own labelling scheme.
+There are [recommended labels](/docs/concepts/overview/working-with-objects/common-labels/#labels) that you can use if you wish.
+{{< /note >}}
+
+## Alternatives
+
+If you plan to create a large number of Jobs, you may find that:
+
+- Even using labels, managing so many Job objects is cumbersome.
+- If you create many Job objects in a batch, you might place high load
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md]: possible edit: line 278, 279: s/Jobs/Job objects/

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 6]
```diff
@@ -1,50 +1,62 @@
...
 
-In this example, we will run multiple Kubernetes Jobs created from
-a common template.  You may want to be familiar with the basic,
-non-parallel, use of [Jobs](/docs/concepts/workloads/controllers/jobs-run-to-completion/) first.
+This task demonstrates running multiple {{< glossary_tooltip text="Jobs" term_id="job" >}}
+based a common template. You can use this approach to process batches of work in
+parallel.
 
+For this example there are only three items: [_apple_, _banana_, _cherry_].
+The sample Jobs process each item simply by printing a string then pausing.
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md]: nit: do you need the square brackets?

**Comment by kbhawkey** [comment]
@sftim I like the change from concept to task. Instead of using Jinja2, sed, kubectl, the task could customize, create, and apply the manifests using the `kustomize` tool or `kubectl -k`. This may show a more straightforward approach.

**Comment by sftim** [comment]
> Instead of using Jinja2, sed, kubectl, the task could customize, create, and apply the manifests using the kustomize tool or kubectl -k. This may show a more straightforward approach.

That's something for another pull request, I think.

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 6 above)
Comment (tengqm) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: Ya. Got to be careful with square brackets ... they are interpreted as out-of-band links in markdown.

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 4 above)
Comment (tengqm) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: Template Job is more accurate in this context, IMHO.

**Comment by sftim** [comment]
/retitle [WIP] Improve Job parallel processing expansion task

**Comment by sftim** [comment]
/retitle Improve Job parallel processing expansion task

**Comment by jimangel** [comment]
/lgtm

@sftim is this ready to merge?

**Comment by kbhawkey** [comment]
@sftim, squash commits?

**Comment by sftim** [comment]
Squashed

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 7]
```diff
@@ -1,51 +1,66 @@
...
+non-parallel, use of [Job](/docs/concepts/jobs/run-to-completion-finite-workloads/).
 
-## Basic Template Expansion
+{{< include "task-tutorial-prereqs.md" >}}
 
-First, download the following template of a job to a file called `job-tmpl.yaml`
+For basic templating you need the command-line utility `sed`.
 
-{{< codenew file="application/job/job-tmpl.yaml" >}}
+To follow the advanced templating example, you need a working install of
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:30]: nit: ... you need a working installation of ... OR
... you first need to install ...

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 8]
```diff
@@ -1,51 +1,66 @@
...
-makes it convenient to operate on all the jobs in this group at once.
-We also put the same label on the pod template so that we can check on all Pods of these Jobs
-with a single command.
-After the job is created, the system will add more labels that distinguish one Job's pods
-from another Job's pods.
-Note that the label key `jobgroup` is not special to Kubernetes. You can pick your own label scheme.
+{{% capture steps %}}
 
-Next, expand the template into multiple files, one for each item to be processed.
+## Create a template Job
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:43]: nit: this still sounds awkward.
Create a template for a Job resource OR
Create a template

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 9]
```diff
@@ -1,51 +1,66 @@
...
+
+First, download the following template of a Job to a file called `job-tmpl.yaml`
+
+{{< codenew file="application/job/job-tmpl.yaml" >}}
 
 ```shell
-# Download job-templ.yaml
+# Use curl to download job-templ.yaml
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:50]: nit: Do you need to download the manifest if the page uses the `codenew` shortcode?

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 10]
```diff
@@ -1,51 +1,66 @@
...
 curl -L -s -O https://k8s.io/examples/application/job/job-tmpl.yaml
+```
+
+This *job template* is not yet a valid Kubernetes {{< glossary_tooltip text="manifest" term_id="manifest" >}}.
+Instead that template is a YAML representation of a Job object with some placeholders
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:55]: nit: Lines 54-55, Could clean up a bit.
The `job-tmpl.yaml` manifest represents a Kubernetes Job object with some placeholders for field values.

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 11]
```diff
@@ -68,11 +83,13 @@ job-banana.yaml
 job-cherry.yaml
 ```
 
-Here, we used `sed` to replace the string `$ITEM` with the loop variable.
-You could use any type of template language (jinja2, erb) or write a program
-to generate the Job objects.
+Here, you used `sed` to replace the string `$ITEM` with the loop variable.
+You could use any type of template language (for example: Jinja2; ERB), or  you
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:87]: nit: Here `sed` replaces
You can use any ... or write a program

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 12]
```diff
@@ -177,46 +209,102 @@ spec:
...
 
-The above template defines parameters for each job object using a list of
-python dicts (lines 1-4).  Then a for loop emits one job yaml object
-for each set of parameters (remaining lines).
-We take advantage of the fact that multiple yaml documents can be concatenated
-with the `---` separator (second to last line).
-.)  We can pipe the output directly to kubectl to
-create the objects.
+The above template defines two parameters for each Job object using a list of
+python dicts (lines 1-4).  Then a for loop emits one Job manifest for each
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:213]: nit: line 213, extra space, `.  Then`. Could remove `Then`. A for loop emits ...

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 13]
```diff
@@ -177,46 +209,102 @@ spec:
...
+set of parameters (remaining lines).
+
+This example relies on a feature of YAML. One YAML file can contain multiple
+documents (Kubernetes manifests, in this case), separated by `---` on a line
+by itself.
+You can pipe the output directly to `kubectl` to create the Job objects.
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:219]: nit:You can pipe the output directly to `kubectl` to create the Jobs.

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 14]
```diff
@@ -177,46 +209,102 @@ spec:
...
+YAML file containing Kubernetes manifests:
 
 ```shell
+# This requires the alias you defined earlier
 cat job.yaml.jinja2 | render_template > jobs.yaml
 ```
 
-Or sent directly to kubectl, like this:
+You can view `jobs.yaml` to verify that `render_template` is working OK.
+
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:236]: nit: Would remove `OK`.
View `jobs.yaml` to verify that the `render_template` script is working [correctly].

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 15]
```diff
@@ -177,46 +209,102 @@ spec:
...
+{{< glossary_tooltip text="labels" term_id="label" >}} that
+distinguish one Job's pods from another Job's pods.
+
+In this example, each Job and its Pod template have a label:
+`jobgroup=jobexample`.
+
+Kubernetes itself pays no attention to labels named `jobgroup`. Setting a label
+for all the Jobs you create from a template makes it convenient to operate on all
+those Jobs at once.
+You can see from the job template that each Pod also gets the same label, so that you
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:283]: nit: line 283 `job template` refers to the original `job-tmpl.yaml`?

File: content/en/docs/tasks/job/parallel-processing-expansion.md [hunk 16]
```diff
@@ -126,30 +144,44 @@ We can use this single command to check on the output of all jobs at once:
...
 
 First, copy and paste the following template of a Job object, into a file called `job.yaml.jinja2`:
 
 
 ```liquid
-{%- set params = [{ "name": "apple", "url": "https://www.orangepippin.com/varieties/apples", },
-                  { "name": "banana", "url": "https://en.wikipedia.org/wiki/Banana", },
-                  { "name": "raspberry", "url": "https://www.raspberrypi.org/" }]
+<<<<<<< HEAD
+{%- set params = [{ "name": "apple", "url": "http://dbpedia.org/resource/Apple", },
```
Comment (kbhawkey) [inline content/en/docs/tasks/job/parallel-processing-expansion.md:182]: @sftim , Double check line 181, Remove

**Comment by sftim** [comment]
/retitle [WIP] Improve Job parallel processing expansion task

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 10 above)
Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md:55]: I prefer my wording to be honest. I really want to make clear that `job-tmpl.yaml` is not intended to be used as a manifest directly.

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 9 above)
Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md:50]: You don't. However, this page is aimed at people who are learning.

I think this one-liner would work too
```shell
for i in apple banana cherry; do printf "---\n" && curl -L -s -q https://k8s.io/examples/application/job/job-tmpl.yaml | sed "s/\$ITEM/$i/"; done | kubectl apply -f -
```

but it's not so easy to learn from or to explain.

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 15 above)
Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md:284]: Revised

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 8 above)
Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: Revised to “Create Jobs based on a template”

File: content/en/docs/tasks/job/parallel-processing-expansion.md (hunk 11 above)
Comment (sftim) [reply content/en/docs/tasks/job/parallel-processing-expansion.md]: I reworded this. Hope it's OK now.

**Comment by sftim** [comment]
/retitle Improve Job parallel processing expansion task

How's this? I hope the revised page is clearly better than the existing one and there's nothing blocking a merge.

**Comment by kbhawkey** [comment]
Thanks @sftim 
/approve
